aced_submission/fhir_store.py

```diff
@@ -0,0 +1,133 @@
+"""Adds, retrieves, and deletes FHIR resources in a FHIR store."""
+import json
+import pathlib
+import sys
+
+import click
+import yaml
+from elasticsearch import Elasticsearch
+
+from aced_submission import NaturalOrderGroup
+from aced_submission.meta_flat_load import read_ndjson, write_bulk_http, DEFAULT_ELASTIC
+
+
+@click.group(cls=NaturalOrderGroup, name='fhir')
+def fhir_store():
+    """Adds, retrieves, and deletes FHIR resources in a FHIR store."""
+    pass
+
+
+def resource_generator(project_id, file_path):
+    """Yield FHIR resources from a ndjson file."""
+    program, project = project_id.split('-')
+    assert program, "program is required"
+    assert project, "project is required"
+
+    for _ in read_ndjson(file_path):
+        _["auth_resource_path"] = f"/programs/{program}/projects/{project}"
+        yield _
+
+def fhir_put(project_id, path, elastic_url) -> list[str]:
+    """Upsert FHIR resources to a FHIR store."""
+    assert project_id.count('-') == 1, f"{project_id} should have a single '-' separating program and project"
+
+    elastic = Elasticsearch([elastic_url], request_timeout=120)
+
+    index = doc_type ='fhir'
+    limit = None
+    logs = []
+    for file_path in pathlib.Path(path).glob('*.ndjson'):
+
+        write_bulk_http(elastic=elastic, index=index, doc_type=doc_type, limit=limit,
+                        generator=resource_generator(project_id, file_path), schema=None)
+
+        logs.append(f"wrote {file_path} to {elastic_url}/{index}")
+
+    return logs
+
+
+def fhir_get(project_id, path, elastic_url) -> list[str]:
+    """Retrieve FHIR resources from FHIR store, write to path/resourceType.ndjson."""
+    assert project_id.count('-') == 1, f"{project_id} should have a single '-' separating program and project"
+    program, project = project_id.split('-')
+    assert program, "program is required"
+    assert project, "project is required"
+
+    elastic = Elasticsearch([elastic_url], request_timeout=120)
+
+    index = doc_type = 'fhir'
+    logs = []
+
+    emitters = {}
+    open_files = []
+
+    pathlib.Path(path).mkdir(parents=True, exist_ok=True)
+
+    def _emitter(_resource_type):
+        """Maintain has of open files."""
+        if resource_type not in emitters:
+            file_path = pathlib.Path(path) / f"{_resource_type}.ndjson"
+            emitters[_resource_type] = file_path.open('w')
+            open_files.append(file_path)
+        return emitters[_resource_type]
+
+    auth_resource_path = f"/programs/{program}/projects/{project}"
+
+    res = elastic.search(index=index, doc_type=doc_type, body={"query": {"match": {"auth_resource_path": auth_resource_path}}})
+    for _ in res['hits']['hits']:
+        resource_type = _['_source']['resourceType']
```

Comment:
Getting a  KeyError: 'resourceType' error on this line when attempting to retrieve FHIR data, could be due to the data I uploaded not being to spec —

```sh
aced_submission fhir get --project_id aced-eimages ./fhir
# ...
#   File "/root/venv/lib/python3.9/site-packages/aced_submission/fhir_store.py", line 78, in fhir_get
#     resource_type = _['_source']['resourceType']
# KeyError: 'resourceType'
```

---

**Successfully adding test FHIR data to existing project:**
```sh
aced_submission fhir put --project_id aced-eimages /studies/Alcoholism/extractions/
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:1 records written
# /root/venv/lib/python3.9/site-packages/elasticsearch/connection/base.py:208: ElasticsearchWarning: the default number of shards will change from [5] to [1] in 7.0.0; if you wish to continue using the default of
#  [5] shards, you must manage this on the create index request or with an index template
#   warnings.warn(message, category=ElasticsearchWarning)
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:1 records written
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:4 records written
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:932 records written
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:237 records written
# INFO:aced_submission.meta_flat_load:Writing bulk to fhir limit None.
# INFO:aced_submission.meta_flat_load:48 records written
# - wrote /studies/Alcoholism/extractions/Task.ndjson to http://10.43.112.102:9200/fhir
# - wrote /studies/Alcoholism/extractions/Specimen.ndjson to http://10.43.112.102:9200/fhir
# - wrote /studies/Alcoholism/extractions/Patient.ndjson to http://10.43.112.102:9200/fhir
# - wrote /studies/Alcoholism/extractions/Observation.ndjson to http://10.43.112.102:9200/fhir
# - wrote /studies/Alcoholism/extractions/DocumentReference.ndjson to http://10.43.112.102:9200/fhir
# - wrote /studies/Alcoholism/extractions/Condition.ndjson to http://10.43.112.102:9200/fhir
```

**Encountering KeyError: 'resourceType' when attempting to get FHIR data:**
```sh
aced_submission fhir get --project_id aced-eimages ./fhir
# Traceback (most recent call last):
#   File "/root/venv/bin/aced_submission", line 8, in <module>
#     sys.exit(cli())
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 829, in __call__
#     return self.main(*args, **kwargs)
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 782, in main
#     rv = self.invoke(ctx)
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
#     return _process_result(sub_ctx.command.invoke(sub_ctx))
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
#     return _process_result(sub_ctx.command.invoke(sub_ctx))
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
#     return ctx.invoke(self.callback, **ctx.params)
#   File "/root/venv/lib/python3.9/site-packages/click/core.py", line 610, in invoke
#     return callback(*args, **kwargs)
#   File "/root/venv/lib/python3.9/site-packages/aced_submission/fhir_store.py", line 129, in _fhir_get
#     logs = fhir_get(project_id, path, elastic_url)
#   File "/root/venv/lib/python3.9/site-packages/aced_submission/fhir_store.py", line 78, in fhir_get
#     resource_type = _['_source']['resourceType']
# KeyError: 'resourceType'
```